target/linux/realtek/files-5.10/drivers/net/dsa/rtl83xx/rtl838x.c

```diff
@@ -1693,6 +1693,77 @@ void rtl838x_set_receive_management_action(int port, rma_ctrl_t type, action_typ
 	}
 }
 
+static void rtl838x_led_init(struct rtl838x_switch_priv *priv)
+{
+	int i;
+	u32 led_mask_l = 0, led_mask_h = 0, pm = 0, v, combo_led_mode;
+	u32 setlen;
+	const __be32 *led_set;
+	char set_name[9];
+	struct device_node *node;
+	const char *mode;
+
+	node = of_find_compatible_node(NULL, NULL, "realtek,rtl8380-leds");
+	if (!node) {
+		pr_info("%s No compatible LED node found\n", __func__);
+		return;
+	}
+
+	for (i = 0; i < priv->cpu_port; i++) {
+		if (!priv->ports[i].phy)
+			continue;
+
+		pm |= BIT(i);
+
+		if (i < 24)
+			led_mask_l |= BIT( priv->ports[i].led_num) - 1;
+		else
+			led_mask_h |= BIT( priv->ports[i].led_num) - 1;
+	}
+	// LED-mask for ports 24-27 must be the same as for lower ports when not used
+	led_mask_h = led_mask_h ? led_mask_h : led_mask_l;
+	sw_w32_mask(0x7, led_mask_l, RTL838X_LED_GLB_CTRL);
+	sw_w32_mask(0x7 << 3, led_mask_h << 3, RTL838X_LED_GLB_CTRL);
+
+	if (of_property_read_u32(node, "combo_led_mode", &combo_led_mode))
+		combo_led_mode = 0;  // Default is no combo leds
+	sw_w32_mask(0x3 << 7, combo_led_mode << 7, RTL838X_LED_GLB_CTRL);
+
+	v = 0;
+	for (i = 0; i < 3; i++) {
+		sprintf(set_name, "led_set%d", i);
+		led_set = of_get_property(node, set_name, &setlen);
+		if (!led_set)
+			break;
+		v |= (be32_to_cpup(led_set) << (i * 5)) | (be32_to_cpup(led_set + 1) << (i * 5 + 15));
```

Comment:
You seem to use "LED set" in the commit message in two ways. Here it is actually interpreted as the index of the i-th per-port LED. This is not the same meaning as on RTL839x and was very confusing to decipher.